File: miney/sky.py
"""
The sky one player sees: its colour, the clouds, sun, moon, stars and the brightness.
"""
from typing import TYPE_CHECKING, Any, Optional

from .exceptions import PlayerOffline
# The same colours as on the screen, so "#ffcc00" means one thing in this library.
from .hud import _color

if TYPE_CHECKING:
    from .luanti import Luanti
    from .player import Player


def _colorspec(value: Any) -> str:
    """
    Turn a colour into the string form Luanti reads everywhere.

    :param value: ``"#rrggbb"``, ``"#rrggbbaa"`` or an integer.
    :return: A ColorSpec string, alpha and all.
    :raises ValueError: For anything that is not a colour.
    """
    number = _color(value)
    return f"#{number:08x}" if number > 0xFFFFFF else f"#{number:06x}"


def _flag(value: Any, name: str) -> bool:
    """
    :param value: What the property was set to.
    :param name: The property's name, for the error message.
    :return: The value, once it is really a bool.
    :raises TypeError: For anything else, ``1`` included.
    """
    if not isinstance(value, bool):
        raise TypeError(
            f"{name} is either True or False, not {value!r}."
        )
    return value



class Sky:
    """
    The sky one player sees, and everything hanging in it.

    Reached through :attr:`Player.sky <miney.Player.sky>` and never created directly::

        >>> p = lt.players.Steve
        >>> p.sky.color = "#101040"
        >>> p.sky.clouds = False
        >>> p.sky.brightness = 0.05
        >>> p.sky.reset()

    .. important::

        **This is what one player sees, not what the world is.** Luanti keeps these
        settings per player and sends them to that one client, so the world stays as
        bright as it was: torches burn the same, plants grow the same, and no mob
        spawns because of anything here.

        .. code-block:: python

            p.sky.brightness = 0.05   # Steve sees night. The world is still noon.
            lt.time_of_day = 0.0      # This is the one that brings the zombies.

    Every setting stays until something puts it back, long after your script has ended -
    :meth:`reset` is the way back. Nothing survives the player leaving the game.

    .. note::

        Some games paint the sky themselves, once a second, and VoxeLibre is one of
        them - that is how rain darkens the day and how the Nether is red. What you set
        here is laid over what that game decided, for this player alone: everybody else
        keeps their weather, and so do the parts of the sky you have not set.
    """

    def __init__(self, luanti: "Luanti", player: "Player"):
        """
        :param luanti: The parent :class:`~miney.Luanti` object.
        :param player: Whose sky this is.
        """
        self.lt = luanti
        #: Whose sky this is.
        self.player = player

    def __repr__(self) -> str:
        return f'<Luanti Sky for "{self.player.name}">'

    def _write(self, part: str, value: Any) -> None:
        """
        Set one part of this player's sky, and hold it there.

        The mod does both halves of that: it makes the engine call, and it remembers
        what was set so that a game which paints its own sky - VoxeLibre repaints every
        player about once a second - has this laid over its own answer for this player.
        Without it a colour was gone before the next line of a script ran.

        Each part keeps the fields it is not given, in the engine and in what the mod
        holds: ``set_sky{clouds = false}`` leaves the colour alone, so one property
        really is one call and nothing has to be read back first.

        :param part: ``"sky"``, ``"sun"``, ``"moon"``, ``"stars"`` or ``"ratio"``.
        :param value: The table the engine call takes, or the number for ``"ratio"``.
        """
        dumps = self.lt.lua.dumps
        self.lt.lua.run(
            f"miney_sky.hold({dumps(self.player.name)}, {dumps(part)}, {dumps(value)})",
            wait=False,
        )

    def _read(self, body: str) -> dict:
        """
        Ask the server about this player's sky.

        :param body: Lua returning a table, with ``online = true`` in it so that an
            answer of nothing can only mean the player is gone.
        :return: That table.
        :raises miney.exceptions.PlayerOffline: If the player is not in the game.
        """
        answer = self.lt.lua.run(
            f"local player = minetest.get_player_by_name("
            f"{self.lt.lua.dumps(self.player.name)}) "
            f"if not player then return nil end "
            f"{body}"
        )
        if not answer:
            raise PlayerOffline(
                f'"{self.player.name}" is not in the game, so there is no sky to look '
                f"at."
            )
        return answer

    @property
    def color(self) -> Optional[str]:
        """
        Get or set one flat colour for the whole sky.

        Setting this replaces Luanti's painted sky - the one that goes blue at noon and
        red at sunset - with a single colour that stays. Set it to ``None`` to get the
        painted sky back.

        The stars, sun and moon keep hanging in it, so a black sky is a night sky
        rather than an empty screen::

            >>> p.sky.color = "#101040"      # deep blue, all day
            >>> p.sky.color
            '#101040'
            >>> p.sky.color = None           # the game's own sky again
            >>> p.sky.color
            None

        :return: ``"#rrggbb"``, or ``None`` while the game paints the sky itself.
        :raises ValueError: If the value is not a colour.
        :raises miney.exceptions.PlayerOffline: When read for a player who left.
        """
        answer = self._read(
            "local s = player:get_sky(true) "
            "return {online = true, type = s.type, color = s.base_color}"
        )
        if answer.get("type") != "plain":
            return None
        rgb = answer.get("color") or {}
        return "#{:02x}{:02x}{:02x}".format(
            int(rgb.get("r", 0)), int(rgb.get("g", 0)), int(rgb.get("b", 0))
        )

    @color.setter
    def color(self, value: Optional[str]) -> None:
        if value is None:
            self._write("sky", {"type": "regular"})
        else:
            self._write("sky", {"type": "plain", "base_color": _colorspec(value)})

    @property
    def clouds(self) -> bool:
        """
        Get or set whether there are clouds.

        >>> p.sky.clouds = False

        :return: ``True`` if the player sees clouds.
        :raises TypeError: If the value is not ``True`` or ``False``.
        :raises miney.exceptions.PlayerOffline: When read for a player who left.
        """
        return bool(
            self._read(
                "local s = player:get_sky(true) "
                "return {online = true, clouds = s.clouds}"
            ).get("clouds")
        )

    @clouds.setter
    def clouds(self, value: bool) -> None:
        self._write("sky", {"clouds": _flag(value, "clouds")})

    @property
    def sun(self) -> bool:
        """
        Get or set whether the sun is in the sky.

        The morning glow on the horizon goes with it. Luanti draws that separately, and
        left on its own it keeps colouring the sunrise for a sun that is not there
        anymore.

        >>> p.sky.sun = False

        :return: ``True`` if the player sees the sun.
        :raises TypeError: If the value is not ``True`` or ``False``.
        :raises miney.exceptions.PlayerOffline: When read for a player who left.
        """
        return bool(
            self._read(
                "local s = player:get_sun() return {online = true, sun = s.visible}"
            ).get("sun")
        )

    @sun.setter
    def sun(self, value: bool) -> None:
        shown = _flag(value, "sun")
        self._write("sun", {"visible": shown, "sunrise_visible": shown})

    @property
    def moon(self) -> bool:
        """
        Get or set whether the moon is in the sky.

        >>> p.sky.moon = False

        :return: ``True`` if the player sees the moon.
        :raises TypeError: If the value is not ``True`` or ``False``.
        :raises miney.exceptions.PlayerOffline: When read for a player who left.
        """
        return bool(
            self._read(
                "local s = player:get_moon() return {online = true, moon = s.visible}"
            ).get("moon")
        )

    @moon.setter
    def moon(self, value: bool) -> None:
        self._write("moon", {"visible": _flag(value, "moon")})

    @property
    def stars(self) -> bool:
        """
        Get or set whether there are stars.

        They are there by day as well, just too faint to see against a bright sky.
        Turn the :attr:`brightness` down and they come out.

        >>> p.sky.stars = True

        :return: ``True`` if the player sees stars.
        :raises TypeError: If the value is not ``True`` or ``False``.
        :raises miney.exceptions.PlayerOffline: When read for a player who left.
        """
        return bool(
            self._read(
                "local s = player:get_stars() return {online = true, stars = s.visible}"
            ).get("stars")
        )

    @stars.setter
    def stars(self, value: bool) -> None:
        self._write("stars", {"visible": _flag(value, "stars")})

    @property
    def brightness(self) -> Optional[float]:
        """
        Get or set how bright the world looks to this player.

        ``0`` is the middle of the night and ``1`` is noon, whatever the clock says.
        ``None`` hands the daylight back to the day and night cycle.

        This is the one that makes a dark sky feel dark: :attr:`color` paints what is
        above the player, this dims everything below it.

        .. code-block:: python

            p.sky.brightness = 0.05     # night, at noon
            p.sky.brightness = None     # back to the time of day

        :return: A number between 0 and 1, or ``None`` while the day decides.
        :raises ValueError: If the value is outside 0 to 1.
        :raises TypeError: If the value is not a number.
        :raises miney.exceptions.PlayerOffline: When read for a player who left.
        """
        ratio = self._read(
            "return {online = true, ratio = player:get_day_night_ratio()}"
        ).get("ratio")
        # The engine keeps this as a 32 bit float, so 0.05 comes back as
        # 0.05000000074505806 and `sky.brightness == 0.05` is False right after setting
        # it. Six decimals is finer than anything an eye can tell apart on a 0 to 1
        # dimmer, and it makes the value that goes out the value that comes back.
        return None if ratio is None else round(float(ratio), 6)

    @brightness.setter
    def brightness(self, value: Optional[float]) -> None:
        if value is None:
            self._write("ratio", None)
            return
        if isinstance(value, bool) or not isinstance(value, (int, float)):
            raise TypeError(
                f"Brightness is a number between 0 (night) and 1 (noon), or None to "
                f"let the day decide. {value!r} is neither."
            )
        if not 0 <= value <= 1:
            raise ValueError(
                f"Brightness has to be between 0 (night) and 1 (noon), not {value!r}."
            )
        self._write("ratio", float(value))

    def reset(self) -> None:
        """
        Put back the sky the game shows by itself.

        Colour, clouds, sun, moon, stars and brightness, all in one go. Worth the last
        line of any script that changed the sky: a player left in an artificial night
        stays in it until they log out.

        .. code-block:: python

            p.sky.color = "#101040"
            p.sky.brightness = 0.05
            # ... the show ...
            p.sky.reset()

        In a game that paints its own sky - VoxeLibre does, for rain, for the Nether,
        for being underwater - this is also the line that hands that job back for this
        player. Nobody else's sky was ever affected, and nobody else's changes here.
        """
        self.lt.lua.run(
            f"miney_sky.release({self.lt.lua.dumps(self.player.name)})",
            wait=False,
        )

File: mod_data/miney/sky.lua
-- The server half of miney/sky.py: one player's sky, and how to keep a game from
-- painting over it.
--
-- Setting a sky is one engine call and would need no mod at all - except that a game
-- may own the sky itself. VoxeLibre does: mcl_weather/skycolor.lua rewrites set_sky,
-- set_sun, set_sun, set_stars and override_day_night_ratio for every player about once
-- a second, which is how rain darkens the day and how the Nether glows red. A colour
-- set from Python was gone before the next line of the script ran.
--
-- That game has a chain of filters it builds each sky from, so Miney adds one at the
-- end: what a script has set for a player is laid over what the game just decided, for
-- that player only. Everybody else keeps their weather, and the fields nobody has
-- claimed - fog, the sky colours behind the clouds - stay the game's.

miney_sky = {}

-- player_name -> {sky = {...}, sun = {...}, moon = {...}, stars = {...}, ratio = n}
-- Only the parts a script has actually set are in here; the rest stays the game's.
local held = {}

-- What Python may name, and the engine call that puts it into the world now. Anything
-- not in here is refused rather than stored, so a typo cannot become a sky nobody can
-- reset.
local PARTS = {
    sky = function(player, value) player:set_sky(value) end,
    sun = function(player, value) player:set_sun(value) end,
    moon = function(player, value) player:set_moon(value) end,
    stars = function(player, value) player:set_stars(value) end,
    ratio = function(player, value) player:override_day_night_ratio(value) end,
}

--- A copy of `base` with `over` written on top of it.
--
-- A copy, because the table came out of the game's own filters and may well be one it
-- keeps and hands to the next player too. Writing into it would give everybody the sky
-- of whoever was updated first.
--
-- @param base (table or nil)
-- @param over (table)
-- @return (table)
local function laid_over(base, over)
    local merged = {}
    for k, v in pairs(base or {}) do merged[k] = v end
    for k, v in pairs(over) do merged[k] = v end
    return merged
end

--- Set one part of a player's sky, and hold it there.
--
-- @param player_name (string)
-- @param part (string) one of the keys of PARTS
-- @param value (table or number or nil) what to pass the engine call
-- @return (boolean) true when it was applied, false when the part is not a part
function miney_sky.hold(player_name, part, value)
    local apply = PARTS[part]
    if not apply then return false end

    local mine = held[player_name]
    if not mine then
        mine = {}
        held[player_name] = mine
    end
    -- set_sky{clouds = false} keeps the colour it was not given, and the same is true
    -- of the others, so a part is merged rather than replaced. The ratio is one number
    -- and has nothing to merge.
    if part == "ratio" then
        mine.ratio = value
    else
        mine[part] = laid_over(mine[part], value or {})
    end

    local player = minetest.get_player_by_name(player_name)
    if not player then return false end
    apply(player, value)
    return true
end

--- Give the sky back: forget what was held and put the engine's own values there.
--
-- @param player_name (string)
-- @return (boolean) true when the player was there to reset
function miney_sky.release(player_name)
    held[player_name] = nil

    local player = minetest.get_player_by_name(player_name)
    if not player then return false end
    player:set_sky()
    player:set_sun()
    player:set_moon()
    player:set_stars()
    player:set_clouds()
    player:override_day_night_ratio()

    -- The game paints again by itself within a few seconds; this only saves the wait.
    if mcl_weather and mcl_weather.skycolor
        and mcl_weather.skycolor.update_player_sky_color then
        mcl_weather.skycolor.update_player_sky_color(player)
    end
    return true
end

--- Lay what a script holds over the sky a game just decided on.
--
-- The shape of `sky_data` is VoxeLibre's, from mcl_weather/skycolor.lua: a `sky` table
-- it passes to set_sky, optional `sun`, `moon` and `stars` tables, and a
-- `day_night_ratio`.
--
-- @param player (ObjectRef)
-- @param sky_data (table) modified in place
local function overlay(player, sky_data)
    local mine = held[player:get_player_name()]
    if not mine then return end

    if mine.sky then sky_data.sky = laid_over(sky_data.sky, mine.sky) end
    if mine.sun then sky_data.sun = laid_over(sky_data.sun, mine.sun) end
    if mine.moon then sky_data.moon = laid_over(sky_data.moon, mine.moon) end
    if mine.stars then sky_data.stars = laid_over(sky_data.stars, mine.stars) end
    if mine.ratio ~= nil then sky_data.day_night_ratio = mine.ratio end
end

-- Last in the chain, so this is the one that decides. after_mods, because the mod that
-- owns the chain may well load after this one.
minetest.register_on_mods_loaded(function()
    if mcl_weather and mcl_weather.skycolor and mcl_weather.skycolor.filters then
        table.insert(mcl_weather.skycolor.filters, overlay)
    end
end)

-- The engine drops a player's sky when they leave anyway, so a leftover entry here
-- would only be a sky that comes back on its own the next time they log in.
minetest.register_on_leaveplayer(function(player)
    held[player:get_player_name()] = nil
end)
